bridge-token/src/lib.rs

```diff
@@ -43,72 +47,86 @@ impl BridgeToken {
+            token: FungibleToken::new(b"t".to_vec()),
+            name: String::default(),
+            symbol: String::default(),
+            reference: String::default(),
+            reference_hash: Base64VecU8(vec![]),
+            decimals: 0,
             paused: Mask::default(),
         }
     }
 
+    pub fn set_metadata(
+        &mut self,
+        name: Option<String>,
+        symbol: Option<String>,
+        reference: Option<String>,
+        reference_hash: Option<Base64VecU8>,
+        decimals: Option<u8>,
+    ) {
+        // Only owner can change the metadata
+        assert_eq!(env::current_account_id(), env::signer_account_id());
+        name.map(|name| self.name = name);
+        symbol.map(|symbol| self.symbol = symbol);
+        reference.map(|reference| self.reference = reference);
+        reference_hash.map(|reference_hash| self.reference_hash = reference_hash);
+        decimals.map(|decimals| self.decimals = decimals);
+    }
+
+    #[payable]
     pub fn mint(&mut self, account_id: AccountId, amount: U128) {
         assert_eq!(
             env::predecessor_account_id(),
             self.controller,
             "Only controller can call mint"
         );
-        self.token.mint(account_id, amount.into());
+
+        self.storage_deposit(Some(account_id.as_str().try_into().unwrap()), None);
+        self.token.internal_deposit(&account_id, amount.into());
     }
 
     #[payable]
     pub fn withdraw(&mut self, amount: U128, recipient: String) -> Promise {
         self.check_not_paused(PAUSE_WITHDRAW);
 
-        self.token
-            .burn(env::predecessor_account_id(), amount.into());
-
-        Promise::new(env::signer_account_id())
-            .transfer(env::attached_deposit())
-            .then(ext_bridge_token_factory::finish_withdraw(
-                amount.into(),
-                recipient,
-                &self.controller,
-                NO_DEPOSIT,
-                env::prepaid_gas() / 2,
-            ))
-    }
-}
+        assert_one_yocto();
+        Promise::new(env::predecessor_account_id()).transfer(1);
 
-#[near_bindgen]
-impl FungibleToken for BridgeToken {
-    #[payable]
-    fn inc_allowance(&mut self, escrow_account_id: String, amount: U128) {
-        self.token.inc_allowance(escrow_account_id, amount.into());
-    }
-
-    #[payable]
-    fn dec_allowance(&mut self, escrow_account_id: String, amount: U128) {
-        self.token.dec_allowance(escrow_account_id, amount.into());
-    }
-
-    #[payable]
-    fn transfer_from(&mut self, owner_id: String, new_owner_id: String, amount: U128) {
         self.token
-            .transfer_from(owner_id, new_owner_id, amount.into());
+            .internal_withdraw(&env::predecessor_account_id(), amount.into());
+
+        ext_bridge_token_factory::finish_withdraw(
+            amount.into(),
+            recipient,
+            &self.controller,
+            NO_DEPOSIT,
+            env::prepaid_gas() / 2,
```

Comment:
This may run out of gas, in which case the tokens will be destroyed.

Reply:
Done https://github.com/near/rainbow-token-connector/pull/39/commits/d0b92affe1243445e3ee54368825a6e6f083772e

Together with fixed gas to schedule `finish_withdraw` / `finish_deposit` / `mint`